oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/web/authentication/OAuth2TokenEndpointAuthenticationSuccessHandlerTests.java

```diff
@@ -0,0 +1,187 @@
+				.extracting("errorCode")
+				.isEqualTo(OAuth2ErrorCodes.SERVER_ERROR);
+
+		verifyNoInteractions(responseConverter);
+	}
+
+	@Test
+	public void onAuthenticationSuccessCustomHttpMessageConverterInjectsAdditionalParams() throws ServletException, IOException {
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		MockHttpServletResponse response = new MockHttpServletResponse();
+
+		Instant issuedAt = Instant.now();
+		Instant expiresAt = issuedAt.plusSeconds(300);
+		OAuth2AccessToken accessToken = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER,
+				"access-token-value", issuedAt, expiresAt, Set.of("scope1"));
+		OAuth2RefreshToken refreshToken = new OAuth2RefreshToken("refresh-token-value", issuedAt);
+		Map<String, Object> additionalParameters = Collections.singletonMap("param1", "value1");
+		Authentication authentication = new OAuth2AccessTokenAuthenticationToken(testClient, clientPrincipal, accessToken, refreshToken, additionalParameters);
+
+		CustomOAuth2AccessTokenResponseHttpMessageConverter customAccessTokenHttpResponseConverter = new CustomOAuth2AccessTokenResponseHttpMessageConverter();
+		authenticationSuccessHandler.setAccessTokenHttpResponseConverter(customAccessTokenHttpResponseConverter);
+		authenticationSuccessHandler.onAuthenticationSuccess(request, response, authentication);
+
+		assertThat(customAccessTokenHttpResponseConverter.getAccessTokenResponse().getAdditionalParameters())
+				.containsExactlyInAnyOrderEntriesOf(Map.of("param1", "value1", "custom_key", "custom_value", "custom_key_2", "custom_value_2"));
+		assertThat(customAccessTokenHttpResponseConverter.getAccessTokenResponse().getAccessToken().getTokenValue()).isEqualTo(accessToken.getTokenValue());
+		assertThat(customAccessTokenHttpResponseConverter.getAccessTokenResponse().getRefreshToken().getTokenValue()).isEqualTo(refreshToken.getTokenValue());
+	}
+
+	/**
+	 * Custom OAuth 2 Access Token Response HTTP message converter demonstrating ability to modify access token
+	 * response before it is sent to the client.
+	 */
+	static class CustomOAuth2AccessTokenResponseHttpMessageConverter extends OAuth2AccessTokenResponseHttpMessageConverter {
+		private OAuth2AccessTokenResponse accessTokenResponse;
+
+		@Override
+		protected boolean supports(Class<?> clazz) {
+			return super.supports(clazz);
+		}
+
+		@Override
+		protected OAuth2AccessTokenResponse readInternal(Class<? extends OAuth2AccessTokenResponse> clazz, HttpInputMessage inputMessage) throws HttpMessageNotReadableException {
+			return super.readInternal(clazz, inputMessage);
+		}
+
+		@Override
+		protected void writeInternal(OAuth2AccessTokenResponse oAuth2AccessTokenResponse, HttpOutputMessage outputMessage) throws HttpMessageNotWritableException {
+			Map<String, Object> allParams = new HashMap<>(oAuth2AccessTokenResponse.getAdditionalParameters());
+			allParams.putAll(Map.of("custom_key", "custom_value", "custom_key_2", "custom_value_2"));
```

Comment:
I'm not sure this approach will work for adding additional parameters. What if the requirement is to obtain specific information from `OAuth2Authorization` and use that for additional parameters? See the sample in this [comment](https://github.com/spring-projects/spring-authorization-server/issues/925#issuecomment-1291178579).

Reply:
I see your point about the authorization information injection use case. I updated the implementation to reflect this. Let me know if it makes sense.

Reply:
@ddubson I took a look at the updated test and I don't think the `HttpMessageConverter` should have the responsibility of looking up `OAuth2Authorization` for additional information. This logic (if needed) should live in the `AuthenticationSuccessHandler`.

I'd like to propose the same pattern we've been using in the following implementations:

- `OAuth2AuthorizationConsentAuthenticationProvider.setAuthorizationConsentCustomizer()`
- `OAuth2AuthorizationCodeRequestAuthenticationProvider.setAuthenticationValidator()`
- `OidcUserInfoAuthenticationProvider.setUserInfoMapper()`

The proposal is to add:

`OAuth2AccessTokenResponseAuthenticationSuccessHandler.setAccessTokenResponseCustomizer(Consumer<OAuth2AccessTokenAuthenticationContext> authenticationContext)`

The `OAuth2AccessTokenAuthenticationContext` would contain `OAuth2AccessTokenAuthenticationToken` and `OAuth2AccessTokenResponse.Builder` for further customization.

Let me know what you think.

Also, I'm off for the holidays and returning Jan 8.
Enjoy your holidays and we'll chat when I'm back.

Reply:
Thanks for the heads up about the context object approach being used. I wasn't aware of it until now and it makes sense. From my cursory glance at the places you mentioned, it seems to be what is needed in this feature as well. Will update this thread when changes are ready for review.

Reply:
@jgrandja I've applied the context object pattern to this PR. Let me know your thoughts.